File: src/leetcode/m97/InterleavingString.java
/*
 * Copyright (c) 2024 THGEEK.COM. All rights reserved.
 */

package leetcode.m97;

/**
 * LeetCode 97
 *
 * @author Tao Hong
 * @version 1.0
 * @since 2024/10/28 12:53
 */
public class InterleavingString {
    public boolean isInterleave(String s1, String s2, String s3) {
        if (s1.length() + s2.length() != s3.length()) {
            return false;
        }

        boolean[][] dp = new boolean[s1.length() + 1][s2.length() + 1];
        dp[0][0] = true;

        for (int i = 0; i <= s1.length(); i++) {
            for (int j = 0; j <= s2.length(); j++) {
                if (i > 0) {
                    dp[i][j] |= dp[i - 1][j] && s1.charAt(i - 1) == s3.charAt(i + j - 1);
                }
                if (j > 0) {
                    dp[i][j] |= dp[i][j - 1] && s2.charAt(j - 1) == s3.charAt(i + j - 1);
                }
            }
        }
        return dp[s1.length()][s2.length()];
    }
}

File: src/leetcode/m97/InterleavingString.go
package m97

// LeetCode 97
func isInterleave(s1 string, s2 string, s3 string) bool {
	m, n, total := len(s1), len(s2), len(s3)
	if m+n != total {
		return false
	}

	dp := make([][]bool, m+1)
	for i := range dp {
		dp[i] = make([]bool, n+1)
	}
	dp[0][0] = true

	for i := 0; i <= m; i++ {
		for j := 0; j <= n; j++ {
			if i > 0 {
				dp[i][j] = dp[i-1][j] && s1[i-1] == s3[i+j-1] || dp[i][j]
			}
			if j > 0 {
				dp[i][j] = dp[i][j-1] && s2[j-1] == s3[i+j-1] || dp[i][j]
			}
		}
	}
	return dp[m][n]
}
